packages/flutter/lib/src/material/tabs.dart

```diff
@@ -760,543 +352,478 @@ class TabBar<T> extends Scrollable implements AppBarBottomWidget {
+  @override
+  void dependenciesChanged() {
+    super.dependenciesChanged();
+    _initTabController();
+    _currentIndex = _controller?.index;
   }
 
   @override
-  void handleStatusChange(AnimationStatus status) {
+  void didUpdateConfig(TabBarView oldConfig) {
+    super.didUpdateConfig(oldConfig);
+    if (config.controller != oldConfig.controller)
+      _initTabController();
+    if (config.children != oldConfig.children && !_warpUnderway)
+      _children = config.children;
   }
 
   @override
-  void handleProgressChange() {
-    if (_selection == null || !_selection.valueIsChanging)
-      return;
-    // The TabBar is driving the TabBarSelection animation.
+  void dispose() {
+    _controller.animation.removeListener(_handleTick);
+    super.dispose();
+  }
 
-    final Animation<double> animation = _selection.animation;
+  void _handleTick() {
+    if (!_controller.indexIsChanging)
+      return; // This widget is driving the controller's animation.
 
-    if (animation.status == AnimationStatus.completed) {
-      _updateItemsAndScrollBehavior();
-      return;
+    if (_controller.index != _currentIndex) {
+      _currentIndex = _controller.index;
+      _warpToCurrentIndex();
     }
+  }
 
-    if (animation.status != AnimationStatus.forward)
-      return;
+  Future<Null> _warpToCurrentIndex() async {
+    if (!mounted)
+      return new Future<Null>.value();
 
-    final int selectedIndex = _selection.index;
-    final int previousSelectedIndex = _selection.previousIndex;
+    final ScrollableState viewport = viewportKey.currentState;
+    if (viewport.scrollOffset == _currentIndex.toDouble())
+      return new Future<Null>.value();
 
-    if (selectedIndex < previousSelectedIndex) {
-      _updateItemsFromChildren(selectedIndex, previousSelectedIndex);
-      scrollTo(new CurveTween(curve: Curves.fastOutSlowIn.flipped).evaluate(new ReverseAnimation(animation)));
-    } else {
-      _updateItemsFromChildren(previousSelectedIndex, selectedIndex);
-      scrollTo(new CurveTween(curve: Curves.fastOutSlowIn).evaluate(animation));
-    }
+    final int previousIndex = _controller.previousIndex;
+    if ((_currentIndex - previousIndex).abs() == 1)
+      return viewport.scrollTo(_currentIndex.toDouble(), duration: kTabScrollDuration);
+
+    assert((_currentIndex - previousIndex).abs() > 1);
+    double initialScroll;
+    setState(() {
+      _warpUnderway = true;
+      _children = new List<Widget>.from(config.children, growable: false);
+      if (_currentIndex > previousIndex) {
+        _children[_currentIndex - 1] = _children[previousIndex];
+        initialScroll = (_currentIndex - 1).toDouble();
+      } else {
+        _children[_currentIndex + 1] = _children[previousIndex];
+        initialScroll = (_currentIndex + 1).toDouble();
+      }
+    });
+    await viewport.scrollTo(initialScroll);
+    await viewport.scrollTo(_currentIndex.toDouble(), duration: kTabScrollDuration);
+    setState(() {
```

Comment:
I have tried to write such a test, but so far I've been unable to dispose the TabBarView in between the awaits. Logically it seems possible, so I've added some defensive `if (mounted) ...` statements.